File: data/update_recipes.js
var { Redis } = require("ioredis");
const fs = require("fs");
var { config } = require("../config.js");
var recipes = require('./resources/recipe_data.js');

const redis = new Redis({
  host: config.REDIS_HOST,
  port: config.REDIS_PORT,
  // password: config.REDIS_PASSWORD,
});



// update all recipes in the database
async function updateRecipes() {
  const id_map = JSON.parse(
    fs.readFileSync("./resources/id_map.json", "utf8")
  );
  // for each recipe in crafting_recipes.js
  let update_count = 0;
  for (var crafter_class in recipes) {
    console.log("Starting " + crafter_class);
    for (var recipe in recipes[crafter_class]) {
      // load the recipe into r
      let r = recipes[crafter_class][recipe];
      
      // get the recipe id
      const recipe_name = r.recipe.trim();
      const recipe_id = id_map[recipe_name];

      // convert the materials and crystals to objects that include the quantity and id 
      for (var mat in r['materials']) {
        const mat_id = id_map[mat];
        const mat_qty = r['materials'][mat];
        r['materials'][mat] = {id: mat_id, qty: mat_qty};
      }

      for (var cry in r['crystals']) {
        const cry_id = id_map[cry];
        const cry_qty = r['crystals'][cry];
        r['crystals'][cry] = {id: cry_id, qty: cry_qty};
      }
      
      // stringify the nested objects
      r['materials'] = JSON.stringify(r['materials'])
      r['crystals'] = JSON.stringify(r['crystals'])

      // set the recipe in redis
      await redis.hset(recipe_id, r);
      update_count++;
    }
  }
  console.log("Updated " + update_count + " recipes");
  return 0;
}
updateRecipes()
  .then(() => {
    console.log("finished");
    process.exit(0);
  })
  .catch((err) => {
    console.log(err);
  });

File: data/update_recipes.py
import redis
import json

r = redis.Redis(host='localhost', port=6379, db=0)

def main():
    # keep track of how many recipes we've skipped
    skipped = 0 
    # Load id_map.json into a variable
    with open('./id_map.json', 'r', encoding='utf-8') as f:
        id_map = json.load(f)
    # Load the recipe data into a variable as well
    with open('./crafting_recipes.json', 'r', encoding='utf-8') as f:
        recipes = json.load(f)
        print(type(recipes))
    # Iterate through the recipes and add the unique ID to each recipe
    for recipe in recipes.items():
        # match the recipe's name to the ID in id_map
        # print(type(recipe))
        try:
            recipe_id = id_map[recipe[0]]
            # add each recipe to Redis using the Unique ID as the key
            # r.hmset(recipe_id, recipe)
            print(f"Added {recipe_id} to Redis.")
        except KeyError:
            print(f"Skipping {recipe[0]} because it has no ID.")
            skipped += 1
            continue
    print(f"Finished adding recipes to Redis. Skipped {skipped} recipes.")

if __name__ == '__main__':
    main()